docs/cvl/expr.md

```diff
@@ -359,3 +362,98 @@ state `s`.
+
+1. The name of a contract imported with a {ref}`using statement <using-stmt>`,
+2. The keyword `nativeBalances`, or
+3. The name of a ghost variable or function
+
+It is an error to use different bases on different sides of the comparison operator, and it is also
+an error to use a comparison basis on one side and not the other.
+The application of the basis restricts the comparison
+to only consider the portion of global state identified by the basis.
+
+If the qualifier is a contract identifier
+imported via `using`, then the comparison operation will only consider the storage fields of that contract. For example:
+
+```cvl
+using MyContract as c;
+using OtherContract as o;
+
+rule compare_state_of_c(env e) {
+   storage init = lastStorage;
+   o.mutateOtherState(e);
+   assert lastStorage[o] == init[o];
+}
+```
+
+will pass verification whereas:
+
+```cvl
+using MyContract as c;
+using OtherContract as o;
+
+rule compare_state_of_c(env e) {
+   storage init = lastStorage;
+   c.mutateContractState(e);
+   assert lastStorage[c] == init[c];
+}
+```
+
+will not. 
+
+```{note}
+Comparing contract's state using this method will **not** compare the balance of the contract between the
+two states.
+```
+
+If the qualifier is the identifier `nativeBalances`, then the account balances
+of all contracts are compared between the two storage states. 
+Finally, if the basis is the name of a ghost function or variable, the values of that
+function/variable are compared between storage states.
+
+Two ghost functions are considered equal if they have the same outputs for all input arguments.
+
+```{warning}
+The default behavior of the Prover on unresolved external calls is to pessimistically havoc contract
+state and balances. This behavior will render most storage comparisons that incorporate such
+state useless. Care should be taken (using {ref}`summarization <summaries>`) to ensure that rules
+that compare storage do not encounter this behavior.
+```
+
+```{warning}
+The grammar admits storage comparisons for both equality and inequality
+that occur arbitrarily nested within expressions. However, support within the Prover for
+these comparisons is primarily aimed at assertions of storage equality, e.g., `assert s1 == s2`.
+Support for storage inequality as well as nesting comparisons within other expressions is considered
+experimental.
+```
+
+```{warning}
+The storage comparison checks for exact equality between every single slot of storage which can
+lead to surprising failures of storage equality assertions. 
+In particular, these failures can happen if an uninitialized storage slot is
+written and then later cleared by Solidity (via the `pop()` function or the `delete` keyword). After the
+clear operation the slot will definitely hold 0, but the prover will not make any assumptions
+about the value of the uninitialized slot which means they can be considered different.
```

Comment:
Is there anything the user can do about this, or are they just out of luck if they want to compare storages on a method that calls `pop`?  I don't see one, but if there is an obvious workaround, it would be worth mentioning.

Reply:
you can turn on an unsafe, underapproximation flag that we are consciously keeping out of the public documentation.